contracts/FlowIDTableStaking.cdc

```diff
@@ -833,41 +875,55 @@ pub contract FlowIDTableStaking {
-
-                        if delRecord.tokensCommitted.balance > 0.0 {
-                            emit DelegatorTokensUnstaked(nodeID: nodeRecord.id, delegatorID: delegator, amount: delRecord.tokensCommitted.balance)
-
-                            // move their committed tokens back to their unstaked tokens
-                            delRecord.tokensUnstaked.deposit(from: <-delRecord.tokensCommitted.withdraw(amount: delRecord.tokensCommitted.balance))
-                        }
-
-                        // Request to unstake all tokens
-                        delRecord.tokensRequestedToUnstake = delRecord.tokensStaked.balance
+                   (nodeRecord.role != UInt8(5) && approvedNodeIDs[nodeID] == nil) {
+                    self.removeAndRefundNodeRecord(nodeRecord: nodeRecord)
+                } else {
+                    //capture whether this is a current node or a pending node
+                    if nodeRecord.initialWeight > 0 {
+                        currentNodeCount[nodeRecord.role] = currentNodeCount[nodeRecord.role]! + 1
+                    } else if candidateNodes!.contains(nodeID) {
+                        pendingNodes[nodeRecord.role]!.append(nodeRecord)
                     }
 
-                    // Clear initial weight because the node is not staked any more
-                    nodeRecord.initialWeight = 0
-                } else {
                     // Set weight to 100
                     // Calculations for node weight will come with a future version of epochs
                     nodeRecord.initialWeight = 100
                 }
             }
+
+            let slotLimits = FlowIDTableStaking.account.borrow<&{UInt8: UInt16}>(from: /storage/flowStakingSlotLimits)
+                ?? panic("Could not load the slot limits object from storage!")
+
+            for nodeRole in currentNodeCount.keys {
+                if currentNodeCount[nodeRole]! >= slotLimits[nodeRole]! {
+                    //if all slots are full, remove and refund all pending nodes
+                    for nodeRecord in pendingNodes[nodeRole]! {
+                        self.removeAndRefundNodeRecord(nodeRecord: nodeRecord)
+                    }
+                } else if currentNodeCount[nodeRole]! + UInt16(pendingNodes[nodeRole]!.length) > slotLimits[nodeRole]! {
+                    //randomly remove pending nodes until the limit is met
+                    let deletionCount: UInt16 = currentNodeCount[nodeRole]! + UInt16(pendingNodes[nodeRole]!.length) - slotLimits[nodeRole]!
+                    var deletionArray: [UInt16] = []
+                    while UInt16(deletionArray.length) < deletionCount {
+                        let selection = UInt16(unsafeRandom() % UInt64(pendingNodes[nodeRole]!.length))
+                        if !deletionArray.contains(selection) {
+                            deletionArray.append(selection)
+                        }
+                    }
```

Comment:
I'm still moderatley concerned about this. 

@joshuahannan Do you know if using a dictionary with a lookup here will actually be faster than this potential O(n^2) lookup?

Using random_shuffle we could still keep this whole thing to O(n)

Reply:
Can elaborate on the solution you're proposing? I'm not sure what using a dictionary with a lookup means here